src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/Dependencies/TransitiveDependencyExtensions.cs

```diff
@@ -120,30 +121,35 @@ public static async Task<bool> IsTransitiveDependencyAsync(this ITransitiveDepen
             return false;
         }
 
+        private static readonly ConcurrentDictionary<IProject, IEnumerable<NuGetReference>> _transitiveDependencyCache = new();
+
         private static IEnumerable<NuGetReference> GetProjectTranstiveDependencies(this IProject project)
         {
-            var set = new HashSet<NuGetReference>();
-            var visited = new HashSet<IProject>();
-
-            var projects = new Queue<IProject>();
-            projects.Enqueue(project);
-
-            do
+            return _transitiveDependencyCache.GetOrAdd(project, p =>
```

Comment:
Yes this is why I wanted y'all to look it over 😁
I wasn't entirely sure if the `Get` here needs to take into account changed dependencies and/or if the `project` object here would be new/diff such that everything would still work.
But I also recognized the crazy-expensive nature of this query and was hoping to see some improvement by doing this.

Reply:
One thing I could do, though, is hold `p.CurrentReferences` alongside the `IEnumerable<NugetReference>` value for the project. If the `.CurrentReferences` set differs when we ask for the transitive dependencies for `p`, recompute.

Reply:
I doubt this specific part is expensive in comparison to actually determining the transitive. This essentially enumerates all the direct packages, and then you pass that to the caching layer. If all the transitive depenendencies are already cached, this will only be the hit of combining stuff and reducing duplicates.

> One thing I could do, though, is hold p.CurrentReferences alongside the IEnumerable<NugetReference> value for the project. 

Maybe. You'd have to enumerate them, and at that point, it may be more complicated than it needs to be.

This makes me think about the `IFeatureCollection.Revision` property so you can identify if there's a difference in the collection. The `List<>` implementation has a property like that, but there's no exposed way to answer the question "Has this collection changed?".

Reply:
> project object here would be new/diff such that everything would still work.

Last I checked, we cached the actual project instance and it is a mutable object

Reply:
> This makes me think about the `IFeatureCollection.Revision` property so you can identify if there's a difference in the collection. The `List<>` implementation has a property like that, but there's no exposed way to answer the question "Has this collection changed?".

Remember when everything had a `.Changed` event? Oh, the days.... 😞

Reply:
New way:
![image](https://user-images.githubusercontent.com/20270743/186231683-4c275541-d8ed-40f7-a153-2b6768dcc221.png)

Original way:
![image](https://user-images.githubusercontent.com/20270743/186233157-5efb0df2-4c8f-405d-94e6-edf3e75c3c3b.png)

I'm comfortable calling this a wash so will revert this change.

Reply:
sounds good. Good to know there's a potential benefit here, but caching is hard to get right (https://www.martinfowler.com/bliki/TwoHardThings.html)